pkg/binpack/minimal_fragmentation.go

```diff
@@ -33,34 +33,67 @@ var MinimalFragmentation = SparkBinPackFunction(func(
 	return SparkBinPack(ctx, driverResources, executorResources, executorCount, driverNodePriorityOrder, executorNodePriorityOrder, nodesSchedulingMetadata, minimalFragmentation)
 })
 
-// minimalFragmentation attempts to pack executors onto as few nodes as possible, ideally a single one
+// minimalFragmentation attempts to pack executors onto as few nodes as possible, ideally a single one.
 // nodePriorityOrder is still used as a guideline, i.e. if an application can fit on multiple nodes, it will pick
-// the first eligible node according to nodePriorityOrder
+// the first eligible node according to nodePriorityOrder. additionally, minimalFragmentation will attempt to avoid
+// mostly empty nodes unless those are required for scheduling or they provide a perfect fit, see a couple examples below.
 //
-// for instance if nodePriorityOrder = [a, b, c, d, e]
-// and we can fit 1 executor on a, 1 executor on b, 3 executors on c, 5 executors on d, 5 executors on e
+// for instance if nodePriorityOrder = [a, b, c, d, e, f]
+// and we can fit 1 executor on a, 1 executor on b, 3 executors on c, 5 executors on d, 5 executors on e, 17 executors on f
 // and executorCount = 11, then we will return:
 // [d, d, d, d, d, e, e, e, e, e, a], true
 //
 // if instead we have executorCount = 6, then we will return:
 // [d, d, d, d, d, a], true
+//
+// if instead we have executorCount = 15, then we will return:
+// [d, d, d, d, d, e, e, e, e, e, c, c, c, a, b], true
+//
+// if instead we have executorCount = 17, then we will return:
+// [f, f, ..., f], true
+//
+// if instead we have executorCount = 19, then we will return:
+// [f, f, ..., f, a, b], true
 func minimalFragmentation(
 	_ context.Context,
 	executorResources *resources.Resources,
 	executorCount int,
 	nodePriorityOrder []string,
 	nodeGroupSchedulingMetadata resources.NodeGroupSchedulingMetadata,
 	reservedResources resources.NodeGroupResources) ([]string, bool) {
-	executorNodes := make([]string, 0, executorCount)
 	if executorCount == 0 {
-		return executorNodes, true
+		return []string{}, true
 	}
 
 	nodeCapacities := capacity.GetNodeCapacities(nodePriorityOrder, nodeGroupSchedulingMetadata, reservedResources, executorResources)
 	nodeCapacities = capacity.FilterOutNodesWithoutCapacity(nodeCapacities)
+	if len(nodeCapacities) == 0 {
+		return nil, false
+	}
+
 	sort.SliceStable(nodeCapacities, func(i, j int) bool {
 		return nodeCapacities[i].Capacity < nodeCapacities[j].Capacity
 	})
+	maxCapacity := nodeCapacities[len(nodeCapacities)-1].Capacity
+	if executorCount < maxCapacity {
+		firstNodeWithMaxCapacityIdx := sort.Search(len(nodeCapacities), func(i int) bool {
+			return nodeCapacities[i].Capacity >= maxCapacity
```

Comment:
a different version of this heuristic could be, search for nodes `>= executorCount + 1`. tradeoff being:

* more aggressively ignoring 'empty' nodes (relative to this set of executors)
* more fragmentation for smaller apps, e.g, trying to scheduler 3 executors on `a(1), b(2), c(2), d(4), e(6)` would end up with `b, b, a` as opposed to `c, c, c` with the PR as it is right now

tbd how this would fare against real world workloads, so probably worth testing both rather than guessing.